File: robocode-mobile/lib/studio/open_in_studio.dart
import 'dart:convert';

/// Languages that open the robotics Studio when no `board` is given. Matches
/// the web `ROBOTICS_LANGS` (only arduino; micropython alone opens the coding
/// studio — it only stays robotics when it carries a robotics `board`, e.g.
/// esp32/pico).
const Set<String> roboticsLangs = {'arduino'};

/// Boards the robotics Studio can open. Matches the web `ROBOTICS_BOARDS`.
/// Used to decide mode=robotics vs coding: a known board always wins over the
/// language heuristic (so micropython on the Pico opens robotics, not coding).
const Set<String> roboticsBoards = {'arduino-uno', 'esp32', 'raspberry-pi-pico'};

/// Max chars of the encoded `diagram` query param (mirrors web
/// `STUDIO_DIAGRAM_MAX`) so the URL stays within browser/WebView limits.
const int studioDiagramMax = 16000;

/// base64url (RFC 4648) of the UTF-8 bytes, padding stripped — byte-identical
/// to the web `encodeStudioCode`, so the same studio URL round-trips.
String encodeStudioCode(String code) {
  return base64Url.encode(utf8.encode(code)).replaceAll('=', '');
}

/// Test-only helper: decode a padded base64url string back to text.
String decodeForTest(String padded) => utf8.decode(base64Url.decode(padded));

/// base64url of the diagram's JSON encoding — byte-identical to the web
/// `encodeStudioDiagram`, so the Studio decodes it the same way regardless of
/// which client (web or mobile) built the link.
String encodeStudioDiagram(Map<String, dynamic> diagram) {
  return base64Url.encode(utf8.encode(jsonEncode(diagram))).replaceAll('=', '');
}

/// Relative Studio path that opens [code] in the right Studio mode.
/// - When [board] is a known robotics board, forces mode=robotics on that
///   board (so e.g. micropython on the Pico opens robotics, not coding).
/// - Else falls back to the language heuristic (only `arduino` is robotics).
/// - When a [diagram] is supplied for a robotics link, appends
///   `&diagram=<base64url(JSON)>`, omitting it past [studioDiagramMax] chars
///   so the URL never blows past practical limits.
String studioHref(
  String language,
  String code, {
  String? board,
  Map<String, dynamic>? diagram,
}) {
  final explicitRobotics = board != null && roboticsBoards.contains(board);
  final isRobotics = explicitRobotics || roboticsLangs.contains(language);
  final params = <String, String>{
    'mode': isRobotics ? 'robotics' : 'coding',
    'lang': language,
    if (isRobotics) 'board': explicitRobotics ? board : 'arduino-uno',
    'code': encodeStudioCode(code),
  };
  if (isRobotics && diagram != null) {
    final enc = encodeStudioDiagram(diagram);
    if (enc.length <= studioDiagramMax) params['diagram'] = enc;
  }
  final query = params.entries
      .map((e) => '${e.key}=${Uri.encodeQueryComponent(e.value)}')
      .join('&');
  return 'studio/new?$query';
}

File: robocode-frontend/src/lib/studio/open-in-studio.ts
// URL helpers for opening a code snippet directly in the RoboCode Studio.
// Snippets are base64url-encoded into the studio URL; the studio page decodes
// them into an unsaved editor buffer (never executed server-side). A baked
// wiring diagram can ride alongside the code so the Studio hydrates the wiring.

export const STUDIO_CODE_MAX = 8000;
export const STUDIO_DIAGRAM_MAX = 16000; // max chars of the encoded `diagram` param
export const ROBOTICS_LANGS = ["arduino"] as const;
/** Boards the robotics Studio can open. Used to decide mode=robotics vs coding. */
export const ROBOTICS_BOARDS: ReadonlySet<string> = new Set(["arduino-uno", "esp32", "raspberry-pi-pico"]);

/** Persisted diagram shape (mirrors the backend `BakedDiagram`). */
export type BakedDiagram = {
  board: string;
  parts: { id: string; type: string; x: number; y: number; rotation?: number; props?: Record<string, string | number | boolean> }[];
  wires: { id: string; from: string; to: string; color?: string }[];
};

function toBase64Url(bytes: Uint8Array): string {
  let bin = "";
  for (const b of bytes) bin += String.fromCharCode(b);
  const b64 = typeof btoa !== "undefined" ? btoa(bin) : Buffer.from(bin, "binary").toString("base64");
  return b64.replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
}

function fromBase64Url(s: string): Uint8Array {
  const b64 = s.replace(/-/g, "+").replace(/_/g, "/");
  const bin = typeof atob !== "undefined" ? atob(b64) : Buffer.from(b64, "base64").toString("binary");
  const out = new Uint8Array(bin.length);
  for (let i = 0; i < bin.length; i++) out[i] = bin.charCodeAt(i);
  return out;
}

export function encodeStudioCode(code: string): string {
  return toBase64Url(new TextEncoder().encode(code));
}

export function decodeStudioCode(param: string): string | null {
  try {
    if (!/^[A-Za-z0-9\-_]+$/.test(param)) return null;
    const text = new TextDecoder().decode(fromBase64Url(param));
    if (text.length === 0 || text.length > STUDIO_CODE_MAX) return null;
    return text;
  } catch {
    return null;
  }
}

/** Encode a diagram to a base64url JSON payload for the studio URL. */
export function encodeStudioDiagram(d: BakedDiagram): string {
  return toBase64Url(new TextEncoder().encode(JSON.stringify(d)));
}

/**
 * Decode + validate a diagram param. Returns null (caller falls back to an empty
 * board) when the payload is malformed, oversize, or doesn't target `boardId`
 * (parts[0] must be the `mcu` board for `boardId`).
 */
export function decodeStudioDiagram(param: string, boardId: string): BakedDiagram | null {
  try {
    if (!/^[A-Za-z0-9\-_]+$/.test(param)) return null;
    if (param.length === 0 || param.length > STUDIO_DIAGRAM_MAX) return null;
    const json = new TextDecoder().decode(fromBase64Url(param));
    const d = JSON.parse(json) as BakedDiagram;
    if (!d || typeof d !== "object" || !Array.isArray(d.parts) || !Array.isArray(d.wires)) return null;
    if (d.board !== boardId) return null;
    const head = d.parts[0];
    if (!head || head.id !== "mcu" || head.type !== `__board__:${boardId}`) return null;
    for (const p of d.parts) {
      if (typeof p.id !== "string" || typeof p.type !== "string") return null;
      if (!Number.isFinite(p.x) || !Number.isFinite(p.y)) return null;
    }
    for (const w of d.wires) {
      if (typeof w.from !== "string" || typeof w.to !== "string") return null;
    }
    return d;
  } catch {
    return null;
  }
}

/**
 * Build the /studio/new URL for a snippet.
 * - When `board` is a known robotics board, force mode=robotics on that board
 *   (so micropython on Pico opens robotics, not coding).
 * - Else fall back to the language heuristic (only `arduino` is robotics).
 * - When a `diagram` is supplied (robotics only), append &diagram=<encoded>,
 *   omitting it past STUDIO_DIAGRAM_MAX so the URL stays within browser limits.
 */
export function studioHref(language: string, code: string, board?: string, diagram?: BakedDiagram): string {
  const params = new URLSearchParams();
  const explicitRobotics = !!board && ROBOTICS_BOARDS.has(board);
  const isRobotics = explicitRobotics || (ROBOTICS_LANGS as readonly string[]).includes(language);
  params.set("mode", isRobotics ? "robotics" : "coding");
  params.set("lang", language);
  if (isRobotics) params.set("board", explicitRobotics ? board! : "arduino-uno");
  params.set("code", encodeStudioCode(code));
  if (isRobotics && diagram) {
    const enc = encodeStudioDiagram(diagram);
    if (enc.length <= STUDIO_DIAGRAM_MAX) {
      params.set("diagram", enc);
    } else if (process.env.NODE_ENV !== "production") {
      console.warn(`[studioHref] diagram param ${enc.length} > ${STUDIO_DIAGRAM_MAX}; omitting wiring.`);
    }
  }
  return `/studio/new?${params.toString()}`;
}
